// cdk8s-cli/lib/plugins/_manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManager = void 0;
const child = __importStar(require("child_process"));
const path = __importStar(require("path"));
const url = __importStar(require("url"));
const fs = __importStar(require("fs-extra"));
const semver = __importStar(require("semver"));
const MODULE_NOT_FOUND_ERROR_CODE = 'MODULE_NOT_FOUND'; // TODO is there a known constant we can use here?
/**
 * A `PluginManager` is responsible for loading (and installing) plugins.
 */
class PluginManager {
    constructor(dir) {
        this.dir = dir;
    }
    load(options) {
        const pkg = this.loadPackage(options.pkg, options.version, options.installEnv ?? {});
        const clazz = pkg.module[options.class];
        if (!clazz) {
            throw new Error(`Unable to locate class '${options.class}' in package '${options.pkg}@${options.version}'. Are you sure you exported it?`);
        }
        return { instance: new clazz(options.properties ?? {}), class: options.class, package: pkg };
    }
    loadPackage(pkg, version, installEnv) {
        if (isRange(version)) {
            // we forbid version ranges because it might give the false impression we will be installing
            // the latest version (which we will not because it would mean contacting NPM on every synth)
            throw new Error(`Unsupported version spec for package ${pkg}: ${version}. Cannot be a range.`);
        }
        const proto = url.parse(pkg).protocol;
        if (proto) {
            // urls are not supported because they don't provide a name with which we can 'require' the module.
            // if needed, we can make the loader smarter and enable this, but not for now.
            throw new Error(`Unsupported package reference: ${pkg}. Can either be an NPM package name, or a local path to a directory`);
        }
        const local = path.isAbsolute(pkg)
            // assume relative paths start with '.' because otherise they
            // are easily confused with npm package names.
            || pkg.startsWith(`.${path.sep}`);
        // local plugins are loaded directly, npm packages are loaded
        // from the plugins directory.
        const modulePath = local ? path.resolve(process.cwd(), pkg) : this.pluginDir(pkg, version);
        const pluginName = local ? modulePath : pkg;
        try {
            return this.require(modulePath, version, pluginName);
        }
        catch (e) {
            if (![MODULE_NOT_FOUND_ERROR_CODE].includes(e.code)) {
                // some unexpected error
                throw e;
            }
            if (local) {
                // if a local plugin is missing, nothing we can do about it
                throw e;
            }
            // otherwise, we install from npm and re-require.
            this.installPackage(pkg, version, installEnv);
            return this.require(modulePath, version, pluginName);
        }
    }
    installPackage(pkg, version, env) {
        const pluginDir = path.join(this.dir, pkg, version);
        fs.mkdirpSync(pluginDir);
        const command = [
            'npm',
            'install', `${pkg}@${version}`,
            '--no-save',
            '--prefix', pluginDir,
        ].join(' ');
        const finalEnv = { ...process.env };
        for (const [key, value] of Object.entries(env)) {
            finalEnv[key] = typeof value === 'string' ? key : JSON.stringify(value);
        }
        console.log(`Installing validation plugin: ${pkg}@${version} (this may take a while the first time around)`);
        child.execSync(command, { stdio: ['ignore', 'pipe', 'pipe'], env: finalEnv });
    }
    require(spec, version, pluginName) {
        const modulePath = require.resolve(spec);
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const module = require(modulePath);
        return { version, pkg: pluginName ?? spec, path: modulePath, module };
    }
    pluginDir(pkg, version) {
        return path.join(this.dir, pkg, version, 'node_modules', pkg);
    }
}
exports.PluginManager = PluginManager;
/**
 * Checks if a given version represents a range, or a pinned version.
 * For example:
 *
 *   - '1.x expands to '>=1.0.0 <2.0.0-0'
 *   - `~1.2' expands to '>=1.2.0 <1.3.0-0'
 *   - '1.2.3' expands to '1.2.3'
 */
function isRange(version) {
    return new semver.Range(version).range !== version;
}
