// cdk8s/lib/cron.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cron = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Represents a cron schedule
 */
class Cron {
    /**
     * Create a cron schedule which runs every minute
     */
    static everyMinute() {
        return Cron.schedule({ minute: '*', hour: '*', day: '*', month: '*', weekDay: '*' });
    }
    /**
     * Create a cron schedule which runs every hour
     */
    static hourly() {
        return Cron.schedule({ minute: '0', hour: '*', day: '*', month: '*', weekDay: '*' });
    }
    /**
     * Create a cron schedule which runs every day at midnight
     */
    static daily() {
        return Cron.schedule({ minute: '0', hour: '0', day: '*', month: '*', weekDay: '*' });
    }
    /**
     * Create a cron schedule which runs every week on Sunday
     */
    static weekly() {
        return Cron.schedule({ minute: '0', hour: '0', day: '*', month: '*', weekDay: '0' });
    }
    /**
     * Create a cron schedule which runs first day of every month
     */
    static monthly() {
        return Cron.schedule({ minute: '0', hour: '0', day: '1', month: '*', weekDay: '*' });
    }
    /**
     * Create a cron schedule which runs first day of January every year
     */
    static annually() {
        return Cron.schedule({ minute: '0', hour: '0', day: '1', month: '1', weekDay: '*' });
    }
    /**
     * Create a custom cron schedule from a set of cron fields
     */
    static schedule(options) {
        return new Cron(options);
    }
    ;
    constructor(cronOptions = {}) {
        const minute = fallback(cronOptions.minute, '*');
        const hour = fallback(cronOptions.hour, '*');
        const month = fallback(cronOptions.month, '*');
        const day = fallback(cronOptions.day, '*');
        const weekDay = fallback(cronOptions.weekDay, '*');
        this.expressionString = `${minute} ${hour} ${day} ${month} ${weekDay}`;
    }
}
exports.Cron = Cron;
_a = JSII_RTTI_SYMBOL_1;
Cron[_a] = { fqn: "cdk8s.Cron", version: "2.70.12" };
function fallback(x, def) {
    return x ?? def;
}
