// cdk8s-cli/lib/upgrades.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLatestVersion = exports.upgradeAvailable = void 0;
const os_1 = require("os");
const path_1 = require("path");
const cdk8s = __importStar(require("cdk8s"));
const fs_extra_1 = require("fs-extra");
const _fs_1 = require("./_fs");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const pkg = require('../package.json');
const CACHE = {
    cacheTtlSec: 60 * 30,
    cacheFile: (0, path_1.join)((0, os_1.homedir)(), '.cdk8s-cli.version'),
};
/**
 * Checks if there is a new version of the CLI available on npm.
 *
 * @returns `undefined` if there is no upgrade available. If there is a new
 * version, returns an object with `latest` (the latest version) and `current`
 * (the current version).
 */
function upgradeAvailable() {
    const latest = getLatestVersion(pkg.name, CACHE);
    if (latest !== pkg.version) {
        return { latest, current: pkg.version };
    }
    else {
        return undefined;
    }
}
exports.upgradeAvailable = upgradeAvailable;
/**
 * Returns the latest version of an npm module. The version is cached for 30min
 * to ~/.cdk8s-cli.version.
 *
 * Never throws.
 *
 * @param module The module name
 * @param options cache options
 */
function getLatestVersion(module, options) {
    let latest = readCache();
    if (!latest) {
        try {
            const info = cdk8s.Yaml.load(`http://registry.npmjs.org/${module}/latest`);
            if (!info || info.length < 1) {
                return undefined;
            }
            latest = info[0].version;
        }
        catch (e) {
            return undefined;
        }
        // cannot determine version, return undefined.
        if (!latest) {
            return undefined;
        }
        writeCache(latest);
    }
    return latest;
    function readCache() {
        try {
            const lastModified = (0, fs_extra_1.statSync)(options.cacheFile).mtime;
            const diff = (new Date().getTime()) - lastModified.getTime();
            if (diff > options.cacheTtlSec * 1000) {
                return undefined;
            }
            return (0, fs_extra_1.readFileSync)(options.cacheFile, 'utf-8').trim();
        }
        catch (e) {
            return undefined;
        }
    }
    function writeCache(version) {
        try {
            _fs_1._fs.writeFileSync(options.cacheFile, version);
        }
        catch (e) {
            return;
        }
    }
}
exports.getLatestVersion = getLatestVersion;
