// cdk8s-plus-28/lib/handler.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Handler = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _action = require("./_action");
/**
 * Defines a specific action that should be taken.
 */
class Handler {
    constructor(tcpSocketOptions, commandOptions, httpGetOptions) {
        this.tcpSocketOptions = tcpSocketOptions;
        this.commandOptions = commandOptions;
        this.httpGetOptions = httpGetOptions;
    }
    /**
     * Defines a handler based on an HTTP GET request to the IP address of the container.
     *
     * @param path The URL path to hit
     * @param options Options
     */
    static fromHttpGet(path, options = {}) {
        return new Handler(undefined, undefined, { path, ...options });
    }
    /**
     * Defines a handler based on a command which is executed within the container.
     *
     * @param command The command to execute
     */
    static fromCommand(command) {
        return new Handler(undefined, { command }, undefined);
    }
    /**
     * Defines a handler based opening a connection to a TCP socket on the container.
     *
     * @param options Options
     */
    static fromTcpSocket(options = {}) {
        return new Handler(options, undefined, undefined);
    }
    /**
     * @internal
     */
    _toKube(cont) {
        const exec = this.commandOptions ? _action.Action.fromCommand(this.commandOptions.command) : undefined;
        const httpGet = this.httpGetOptions ? _action.Action.fromHttpGet(cont, this.httpGetOptions.path, this.httpGetOptions) : undefined;
        const tcpSocket = this.tcpSocketOptions ? _action.Action.fromTcpSocket(cont, this.tcpSocketOptions) : undefined;
        return { exec, httpGet, tcpSocket };
    }
}
exports.Handler = Handler;
_a = JSII_RTTI_SYMBOL_1;
Handler[_a] = { fqn: "cdk8s-plus-28.Handler", version: "2.5.6" };
